### PR TITLE
add more ioctl integration

### PR DESCRIPTION
**Pull request by P4N74**

### CODE REVIEW

**Comment by Maijin** [comment]
@P4N74  Add tests for those commands in r2r please

**Comment by radare** [comment]
Tests for a kernel module? I dont think travis will allow that

**Comment by Maijin** [comment]
yeah, didn't realized that was kernel stuff sorry :P But if it's possible maybe in t.archos stuff

**Comment by radare** [comment]
imho this is hard to test, because it depends on so many parameters. also the current code uses eprintf and printf when it should be using io->cb_printf(). about the r_print_hexdump().. we should have access to the RPrint in RCore to reuse the same settings. usually the io plugins don't have access to this, and it would be better to try to avoid as much as possible this lateral communication. so the solution for this would be to make this plugin hold a new instance of RPrint without colors or anything, just to set the cb_printf callback in there instead of the default `printf`. this can be done in another pr

File: libr/io/p/io_r2k.c [hunk 1]
```diff
@@ -254,24 +256,56 @@ struct r2k_data {
...
+#define IOCTL_READ_PHYSICAL_ADDR  _IOR (R2_TYPE, READ_PHYSICAL_ADDR, sizeof (struct r2k_data))
+#define IOCTL_WRITE_PHYSICAL_ADDR _IOR (R2_TYPE, WRITE_PHYSICAL_ADDR, sizeof (struct r2k_data))
+#define IOCTL_GET_PROC_MAPS       _IOR (R2_TYPE, GET_PROC_MAPS, sizeof (struct r2k_data))
+#define IOCTL_GET_KERNEL_MAP      _IOR (R2_TYPE, GET_KERNEL_MAP, sizeof (struct r2k_data))
+
+static ut64 getvalue (const char *buf, int pos) {
+	//TODO: Add checks if the particular position exist or not.
+	int i;
+	ut64 ret;
+	for (i = 0; i < pos; i++) {
```
Comment (lowlyw) [inline libr/io/p/io_r2k.c]: can this part use `getargpos`?

File: libr/io/p/io_r2k.c [hunk 2]
```diff
@@ -313,20 +347,121 @@ static int WriteKernelMemory_linux (RIODesc *iodesc, ut64 address, const ut8 *bu
...
+	buf = r_str_ichr ((char *)buf, ' ');
+
+	switch(*buf) {
+	case 'r':
+		{
+			if (buf[1] == 'l') {
+				//read linear address
+				//=! rl addr len
+				pid = 0;
+				addr = getvalue (buf, 1);
```
Comment (lowlyw) [inline libr/io/p/io_r2k.c]: fix indentation

File: libr/io/p/io_r2k.c [hunk 3]
```diff
@@ -313,20 +347,121 @@ static int WriteKernelMemory_linux (RIODesc *iodesc, ut64 address, const ut8 *bu
...
+				len = getvalue (buf, 2);
+			    ioctl_n = IOCTL_READ_KERNEL_MEMORY;
+			} else if (buf[1] == 'p') {
+				//read process address
+				//=! rp pid address len
+				pid = getvalue (buf, 1);
+				addr = getvalue (buf, 2);
```
Comment (lowlyw) [inline libr/io/p/io_r2k.c]: fix indentation for thsi block too

File: libr/io/p/io_r2k.c [hunk 4]
```diff
@@ -313,20 +347,121 @@ static int WriteKernelMemory_linux (RIODesc *iodesc, ut64 address, const ut8 *bu
...
+				len = getvalue (buf, 3);
+			    ioctl_n = IOCTL_READ_PROCESS_ADDR;
+			} else if (buf[1] == 'P') {
+				//read physical address
+				//=! rP address len
+				pid = 0;
+				addr = getvalue (buf, 1);
```
Comment (lowlyw) [inline libr/io/p/io_r2k.c]: eek, more indentation!

Comment (P4N74) [inline libr/io/p/io_r2k.c]: Yup. I realized it earlier. Haven't updated the PR yet :). Thanks for review though.

File: libr/io/p/io_r2k.c (hunk 1 above)
Comment (P4N74) [inline libr/io/p/io_r2k.c]: Sounds good. Sure, I will do it and update the pr :)

**Comment by radare** [comment]
Even in archos is much harder to test than the debugger

**Comment by P4N74** [comment]
@radare added the `io->cb_printf()` function. Should I include the check `io && io->cb_printf` before calling the function? Or it will be guaranteed that those won't be NULL.

**Comment by radare** [comment]
No need to check that. Io should always have valid poinyers for this. Did u fixed the spacing issues?

**Comment by Maijin** [comment]
image

CLANG/OSX

**Comment by P4N74** [comment]
@radare You mean the indentation issue that lowlyw suggested? In that case, yes. :)

**Comment by P4N74** [comment]
@Maijin thanks :) Updated the pr :)

**Pull request merged**